pkg/optimization/device_perf_linux.go

```diff
@@ -43,79 +45,105 @@ func (deviceHelper *DeviceHelper) DiskSupportsPerfOptimization(diskPerfProfile,
 	}
 
-	err = echoToFile(queueDepth, filepath.Join(blockDeviceRootPath, deviceName, "device/queue_depth"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/queue_depth for device %s. Error: %v", deviceName, err)
-	}
+	deviceSettings = make(map[string]string)
+	deviceSettings[filepath.Join(deviceRoot, "queue/max_sectors_kb")] = maxSectorsKb
+	deviceSettings[filepath.Join(deviceRoot, "queue/scheduler")] = scheduler
+	deviceSettings[filepath.Join(deviceRoot, "device/queue_depth")] = queueDepth
+	deviceSettings[filepath.Join(deviceRoot, "queue/nr_requests")] = nrRequests
+	deviceSettings[filepath.Join(deviceRoot, "queue/read_ahead_kb")] = "8"
 
-	err = echoToFile(nrRequests, filepath.Join(blockDeviceRootPath, deviceName, "queue/nr_requests"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/nr_requests for device %s. Error: %v", deviceName, err)
-	}
+	return deviceSettings, nil
+}
 
-	err = echoToFile(readAheadKb, filepath.Join(blockDeviceRootPath, deviceName, "queue/read_ahead_kb"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/read_ahead_kb for device %s. Error: %v", deviceName, err)
+func getDeviceSettingsForAdvancedProfile(deviceRoot string, deviceSettingsFromCtx map[string]string) (deviceSettings map[string]string, err error) {
+	klog.V(2).Infof("getDeviceSettingsForAdvancedProfile: Getting settings for deviceRoot %s deviceSettingsFromCtx %v",
+		deviceRoot,
+		deviceSettingsFromCtx)
+	deviceSettings = make(map[string]string)
+	for setting, value := range deviceSettingsFromCtx {
+		deviceSettings[filepath.Join(deviceRoot, setting)] = value
 	}
 
-	err = echoToFile("0", filepath.Join(blockDeviceRootPath, deviceName, "queue/wbt_lat_usec"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/wbt_lat_usec for device %s. Error: %v", deviceName, err)
+	return deviceSettings, nil
+}
+
+func updateDeviceSettings(deviceRoot string, deviceSettings map[string]string) (err error) {
+	if len(deviceSettings) == 0 {
+		return fmt.Errorf("updateDeviceSettings: No deviceSettings passed")
 	}
 
-	err = echoToFile("0", filepath.Join(blockDeviceRootPath, deviceName, "queue/rotational"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/rotational for device %s. Error: %v", deviceName, err)
+	for setting, value := range deviceSettings {
+		// Make sure all setting paths are absolute
+		absSetting, err := filepath.Abs(setting)
+		if err != nil {
+			return fmt.Errorf("updateDeviceSettings: Failed to get absolute path. Could not apply setting %s for device %s",
+				setting,
+				deviceRoot)
+		}
+
+		// make sure all settings are under the device root
+		if !strings.HasPrefix(strings.ToLower(absSetting), strings.ToLower(deviceRoot)+"/") {
```

Comment:
Linux file systems are case-sensitive, so we should not make this a case-insensitive comparison. Consider using [`filepath.Match`](https://pkg.go.dev/path/filepath#Match) instead, e.g.

```go
if matches, _ := filepath.Match(filepath.Join(deviceRoot, "*"), absSetting); matches {
   ...
}
```

Because each settings is validated just before it is applied:
1. An incomplete set of settings will be applied if a setting after the first was good
2. Any volume created with a bad set of settings will never be usable (parameters will be stored in the PV)

Consider having a separate validation function to validate the parameters as a group that can be called here and when the volume is created to prevent both of these situations.

Reply:
Thanks for this feedback Ed. So I looked at the implementation for match and HasPrefix both and it looks like Match(* also wont work for us, because it can match only up to one separator. I went ahead and did a case sensitive prefix match using strings.HasPrefix.

I also went ahead and implemented a validation logic in CreateVolume using a dummy device.